cmd/gitops-pusher/gitops-pusher.go

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Command gitops-pusher allows users to use a GitOps flow for managing Tailscale ACLs.
+//
+// See README.md for more details.
+package main
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+)
+
+var (
+	policyFname = flag.String("policy-file", "./policy.hujson", "filename for policy file")
+	timeout     = flag.Duration("timeout", 5*time.Minute, "timeout for the entire CI run")
+)
+
+func main() {
+	flag.Parse()
+	ctx, cancel := context.WithTimeout(context.Background(), *timeout)
+	defer cancel()
+
+	tailnet, ok := os.LookupEnv("TS_TAILNET")
+	if !ok {
+		log.Fatal("set envvar TS_TAILNET to your tailnet's name")
+	}
+	apiKey, ok := os.LookupEnv("TS_API_KEY")
+	if !ok {
+		log.Fatal("set envvar TS_API_KEY to your Tailscale API key")
+	}
+
+	switch flag.Arg(0) {
+	case "apply":
+		controlEtag, err := getACLETag(ctx, tailnet, apiKey)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		localEtag, err := sumFile(*policyFname)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		log.Printf("control: %s", controlEtag)
+		log.Printf("local:   %s", localEtag)
+
+		if controlEtag == localEtag {
+			log.Println("no update needed, doing nothing")
+			os.Exit(0)
+		}
+
+		if err := applyNewACL(ctx, tailnet, apiKey, *policyFname, controlEtag); err != nil {
+			log.Fatal(err)
+		}
+
+	case "test":
+		controlEtag, err := getACLETag(ctx, tailnet, apiKey)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		localEtag, err := sumFile(*policyFname)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		log.Printf("control: %s", controlEtag)
+		log.Printf("local:   %s", localEtag)
+
+		if controlEtag == localEtag {
+			log.Println("no updates found, doing nothing")
+			os.Exit(0)
+		}
+
+		if err := testNewACLs(ctx, tailnet, apiKey, *policyFname); err != nil {
+			log.Fatal(err)
+		}
+	default:
+		log.Fatalf("usage: %s [options] <test|apply>", os.Args[0])
+	}
+}
+
+func sumFile(fname string) (string, error) {
+	fin, err := os.Open(fname)
+	if err != nil {
+		return "", err
+	}
+	defer fin.Close()
+
+	h := sha256.New()
+	_, err = io.Copy(h, fin)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("\"%x\"", h.Sum(nil)), nil
```

Comment:
With the Go hash type, you collect data into the Hash value and then call `Sum` when you want to get the final checksum. This is slightly weird as written because the ETag header value is also quoted. I did it like this to avoid having to shuck or parse the string.